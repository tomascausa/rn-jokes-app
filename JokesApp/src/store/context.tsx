/* eslint-disable react-hooks/exhaustive-deps */
import React, {FC, useState, createContext, useEffect} from 'react';
import {Joke} from '../types/Joke';
import {JokeSchema} from '../realm/schema';
import Realm from 'realm';

interface Props {
    jokes: Joke[];
    addJoke: (joke: Joke) => void;
    removeJoke: (id: string) => void;
}

export const Context = createContext<Props>({
    jokes: [],
    addJoke: (joke: Joke) => {},
    removeJoke: (id: string) => {},
});

const ContextProvider: FC = ({children}) => {
    const [realm, setRealm] = useState<Realm>();
    const [jokes, setJokes] = useState<Joke[]>([]);

    useEffect(() => {
        const openRealm = async () => {
            const realmInstance = await Realm.open({schema: [JokeSchema]});

            if (realmInstance != null && !realmInstance.isClosed) {
                setRealm(realmInstance);

                const response = realmInstance
                    .objects<Joke>('Joke')
                    .map(value => {
                        return value.toJSON() as Joke;
                    });

                setJokes(response);
            }
        };

        openRealm();

        return () => {
            if (!!realm && !realm.isClosed) {
                realm.close();
            }
        };
    }, []);

    const addJoke = (joke: Joke) => {
        if (!jokes.includes(joke)) {
            realmAddJoke(joke);
            setJokes((items: Joke[]) => [...items, joke]);
        }
    };

    const removeJoke = (id: string) => {
        realmRemoveJoke(id);
        setJokes((items: Joke[]) =>
            items.filter((joke: Joke) => joke.id !== id),
        );
    };

    const realmAddJoke = (joke: Joke) => {
        if (!!realm && !realm.isClosed) {
            realm.write(() => {
                realm.create('Joke', joke);
            });
        }
    };

    const realmRemoveJoke = (id: string) => {
        if (!!realm && !realm.isClosed) {
            realm.write(() => {
                const joke = realm.objectForPrimaryKey<Joke>('Joke', id);
                realm.delete(joke);
            });
        }
    };

    return (
        <Context.Provider
            value={{
                jokes,
                removeJoke,
                addJoke,
            }}>
            {children}
        </Context.Provider>
    );
};

export default ContextProvider;
