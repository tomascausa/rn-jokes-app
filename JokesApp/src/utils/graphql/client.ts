import {ApolloClient, InMemoryCache} from '@apollo/client';

export const client = new ApolloClient({
    uri: 'https://icanhazdadjoke.com/graphql',
    cache: new InMemoryCache(),
});
