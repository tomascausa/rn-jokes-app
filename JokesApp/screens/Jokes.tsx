import React, {useContext, useState} from 'react';
import {useNavigation} from '@react-navigation/native';
import {
    Text,
    View,
    StyleSheet,
    SafeAreaView,
    Pressable,
    ActivityIndicator,
} from 'react-native';
import {FontAwesomeIcon} from '@fortawesome/react-native-fontawesome';
import {faHeart} from '@fortawesome/free-solid-svg-icons';
import {Context} from '../store/context';
import {Joke} from '../types/Joke';
import Button from '../components/UI/Button';
import JokeCard from '../components/UI/JokeCard';
import Carousel from '../components/Carousel/Carousel';
import useJokes from '../hooks/useJokes';

const Jokes: React.FC = () => {
    const navigation = useNavigation();
    const context = useContext(Context);
    const {jokes, loading} = useJokes();
    const [currentIndex, setCurrentIndex] = useState(0);

    const handleIndexUpdate = (index: number) => {
        setCurrentIndex(index);
    };

    const handleSaveJoke = () => {
        context.addJoke(jokes[currentIndex]);
    };

    return (
        <SafeAreaView style={styles.container}>
            <View style={styles.header}>
                <Text style={styles.headerText}>
                    Saved: {context.jokes.length}
                </Text>
                <Pressable
                    style={styles.headerHeart}
                    onPress={() => navigation.navigate('Saved')}>
                    <FontAwesomeIcon icon={faHeart} color={'red'} size={28} />
                </Pressable>
            </View>
            <View style={styles.titleContainer}>
                <Text style={styles.title}>
                    Things you can say to annoy designers.
                </Text>
            </View>
            <View style={styles.carouselContainer}>
                {loading ? (
                    <View style={styles.loadingContainer}>
                        <ActivityIndicator size="large" />
                    </View>
                ) : (
                    <Carousel
                        items={jokes}
                        renderItem={(scrollX: any, {item}: {item: Joke}) => (
                            <JokeCard data={item} />
                        )}
                        onIndexUpdate={handleIndexUpdate}
                    />
                )}
            </View>
            <View style={styles.buttonContainer}>
                <Button onPress={() => handleSaveJoke()}>Save</Button>
            </View>
        </SafeAreaView>
    );
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
        flexDirection: 'column',
        marginTop: 60,
        paddingVertical: 40,
        marginHorizontal: 5,
    },
    header: {
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'flex-end',
        paddingHorizontal: 20,
    },
    headerText: {
        fontFamily: 'Roboto-Medium',
        fontSize: 18,
        fontWeight: '500',
    },
    headerHeart: {
        marginLeft: 8,
    },
    titleContainer: {
        marginTop: 15,
        paddingHorizontal: 25,
    },
    title: {
        fontFamily: 'Roboto-Bold',
        fontSize: 36,
        fontWeight: '700',
        textAlign: 'center',
        lineHeight: 45,
    },
    loadingContainer: {
        flex: 1,
        alignItems: 'center',
        justifyContent: 'center',
    },
    carouselContainer: {
        flex: 1,
        marginTop: 40,
    },
    buttonContainer: {
        marginTop: 40,
        paddingHorizontal: 20,
    },
});

export default Jokes;
