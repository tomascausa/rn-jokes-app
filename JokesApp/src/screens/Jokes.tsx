import React, {useContext, useState} from 'react';
import {useNavigation} from '@react-navigation/native';
import {
    Text,
    View,
    StyleSheet,
    SafeAreaView,
    Pressable,
    ActivityIndicator,
    Dimensions,
} from 'react-native';
import {FontAwesomeIcon} from '@fortawesome/react-native-fontawesome';
import {faHeart} from '@fortawesome/free-solid-svg-icons';
import {Context} from '../store/context';
import {Joke} from '../types/Joke';
import {GlobalStyles} from '../constants/styles';
import Button from '../components/UI/Button';
import JokeCard from '../components/UI/JokeCard';
import Carousel from '../components/Carousel/Carousel';
import useJokes from '../hooks/useJokes';

const {width} = Dimensions.get('window');

const Jokes: React.FC = () => {
    const navigation = useNavigation();
    const context = useContext(Context);
    const {jokes, loading, fetchMore} = useJokes();
    const [currentIndex, setCurrentIndex] = useState(0);

    const handleIndexUpdate = (index: number) => {
        setCurrentIndex(index);
    };

    const handleSaveJoke = () => {
        context.addJoke(jokes[currentIndex]);
    };

    const handleFetchMore = () => {
        fetchMore();
    };

    return (
        <SafeAreaView style={styles.container}>
            <View style={styles.header}>
                <Text style={styles.headerText}>
                    Saved: {context.jokes.length}
                </Text>
                <Pressable
                    style={styles.headerHeart}
                    onPress={() => navigation.navigate('Saved')}>
                    <FontAwesomeIcon icon={faHeart} color={'red'} size={28} />
                </Pressable>
            </View>
            <View style={styles.titleContainer}>
                <Text style={styles.title}>
                    Things you can say to annoy designers.
                </Text>
            </View>
            <View style={styles.carouselContainer}>
                {loading ? (
                    <View style={styles.loadingContainer}>
                        <ActivityIndicator size="large" />
                    </View>
                ) : (
                    <Carousel
                        items={jokes}
                        renderItem={(scrollX: any, {item}: {item: Joke}) => (
                            <JokeCard
                                data={item}
                                style={styles.cardContainer}
                                titleStyle={styles.cardTitle}
                                authorStyle={styles.cardAuthor}
                            />
                        )}
                        onIndexUpdate={handleIndexUpdate}
                        onFetchMore={handleFetchMore}
                    />
                )}
            </View>
            <View style={styles.buttonContainer}>
                <Button onPress={() => handleSaveJoke()}>Save</Button>
            </View>
        </SafeAreaView>
    );
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
        flexDirection: 'column',
        marginTop: 60,
        paddingVertical: 40,
        marginHorizontal: 5,
    },
    header: {
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'flex-end',
        paddingHorizontal: 20,
    },
    headerText: {
        fontFamily: 'Roboto-Medium',
        fontSize: 18,
        fontWeight: '500',
    },
    headerHeart: {
        marginLeft: 8,
    },
    titleContainer: {
        marginTop: 15,
        paddingHorizontal: 25,
    },
    title: {
        fontFamily: 'Roboto-Bold',
        fontSize: 36,
        fontWeight: '700',
        textAlign: 'center',
        lineHeight: 45,
    },
    loadingContainer: {
        flex: 1,
        alignItems: 'center',
        justifyContent: 'center',
    },
    carouselContainer: {
        flex: 1,
        marginTop: 40,
    },
    cardContainer: {
        width: width - 50,
        marginHorizontal: 20,
        marginBottom: 30,
        paddingTop: 60,
        paddingBottom: 30,
        paddingHorizontal: 20,
        shadowColor: GlobalStyles.colors.primary,
        shadowOffset: {
            width: 0,
            height: 15,
        },
        shadowOpacity: 0.3,
        shadowRadius: 10,
        elevation: 15,
        borderRadius: 8,
        backgroundColor: GlobalStyles.colors.primary,
    },
    cardTitle: {
        fontSize: 30,
        lineHeight: 42,
    },
    cardAuthor: {
        fontSize: 25,
    },
    buttonContainer: {
        marginTop: 40,
        paddingHorizontal: 20,
    },
});

export default Jokes;
