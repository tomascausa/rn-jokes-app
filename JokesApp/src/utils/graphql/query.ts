import {gql} from '@apollo/client';

export const JOKES_QUERY = gql`
    query {
        joke {
            id
            joke
            permalink
        }
    }
`;
