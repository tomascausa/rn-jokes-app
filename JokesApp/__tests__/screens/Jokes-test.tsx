import React from 'react';
import {MockedProvider} from '@apollo/client/testing';
import {JOKES_QUERY} from '../../src/utils/graphql/query';
import {fireEvent, render} from '@testing-library/react-native';
import {Context} from '../../src/store/context';
import {NavigationContainer} from '@react-navigation/native';
import renderer from 'react-test-renderer';
import Jokes from '../../src/screens/Jokes';

const mocks: any[] = [
    {
        request: {
            query: JOKES_QUERY,
        },
        result: {
            data: {
                joke: {
                    id: '1',
                    joke: 'Test',
                    permalink: 'https://example.com',
                },
            },
        },
    },
];

jest.mock('@fortawesome/react-native-fontawesome', () => ({
    FontAwesomeIcon: '',
}));

describe('<Jokes />', () => {
    test('renders correctly', () => {
        const component = (
            <MockedProvider mocks={mocks}>
                <NavigationContainer>
                    <Jokes />
                </NavigationContainer>
            </MockedProvider>
        );
        const json = renderer.create(component).toJSON();
        expect(json).toMatchSnapshot();
    });

    test('addJoke context method gets called when clicking Save button', async () => {
        const mockAddJoke = jest.fn();
        const component = (
            <MockedProvider mocks={mocks}>
                <Context.Provider
                    value={{
                        jokes: [],
                        removeJoke: jest.fn(),
                        addJoke: mockAddJoke,
                    }}>
                    <NavigationContainer>
                        <Jokes />
                    </NavigationContainer>
                </Context.Provider>
            </MockedProvider>
        );
        const {getByText} = await render(component);
        const toClick = await getByText('Save');

        fireEvent.press(toClick);

        expect(mockAddJoke).toHaveBeenCalledTimes(1);
    });
});
