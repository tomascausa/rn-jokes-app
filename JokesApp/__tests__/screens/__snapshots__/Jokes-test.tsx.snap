// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Jokes /> renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
      "flexDirection": "column",
      "marginHorizontal": 5,
      "marginTop": 60,
      "paddingVertical": 40,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "flex-end",
        "paddingHorizontal": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "fontFamily": "Roboto-Medium",
          "fontSize": 18,
          "fontWeight": "500",
        }
      }
    >
      Saved: 
      0
    </Text>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginLeft": 8,
        }
      }
    >
      <
        color="red"
        icon={
          Object {
            "icon": Array [
              512,
              512,
              Array [
                128153,
                128154,
                128155,
                128156,
                128420,
                129293,
                129294,
                129505,
                10084,
                61578,
                9829,
              ],
              "f004",
              "M0 190.9V185.1C0 115.2 50.52 55.58 119.4 44.1C164.1 36.51 211.4 51.37 244 84.02L256 96L267.1 84.02C300.6 51.37 347 36.51 392.6 44.1C461.5 55.58 512 115.2 512 185.1V190.9C512 232.4 494.8 272.1 464.4 300.4L283.7 469.1C276.2 476.1 266.3 480 256 480C245.7 480 235.8 476.1 228.3 469.1L47.59 300.4C17.23 272.1 .0003 232.4 .0003 190.9L0 190.9z",
            ],
            "iconName": "heart",
            "prefix": "fas",
          }
        }
        size={28}
      />
    </View>
  </View>
  <View
    style={
      Object {
        "marginTop": 15,
        "paddingHorizontal": 25,
      }
    }
  >
    <Text
      style={
        Object {
          "fontFamily": "Roboto-Bold",
          "fontSize": 36,
          "fontWeight": "700",
          "lineHeight": 45,
          "textAlign": "center",
        }
      }
    >
      Things you can say to annoy designers.
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "marginTop": 40,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flex": 1,
          },
          null,
        ]
      }
    >
      <RCTScrollView
        collapsable={false}
        data={Array []}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={true}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onEndReached={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        onViewableItemsChanged={[Function]}
        pagingEnabled={true}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={64}
        showsHorizontalScrollIndicator={false}
        stickyHeaderIndices={Array []}
        style={Object {}}
        testID="flatList"
        viewabilityConfig={
          Object {
            "itemVisiblePercentThreshold": 90,
            "minimumViewTime": 5,
            "waitForInteraction": true,
          }
        }
        viewabilityConfigCallbackPairs={
          Array [
            Object {
              "onViewableItemsChanged": [Function],
              "viewabilityConfig": Object {
                "itemVisiblePercentThreshold": 90,
                "minimumViewTime": 5,
                "waitForInteraction": true,
              },
            },
          ]
        }
      >
        <View />
      </RCTScrollView>
      <View
        collapsable={false}
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
            "marginTop": 10,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "black",
              "borderRadius": 8,
              "height": 5,
              "marginHorizontal": 8,
              "opacity": 1,
              "width": 50,
            }
          }
        />
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "black",
              "borderRadius": 8,
              "height": 5,
              "marginHorizontal": 8,
              "opacity": 0.3,
              "width": 50,
            }
          }
        />
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "black",
              "borderRadius": 8,
              "height": 5,
              "marginHorizontal": 8,
              "opacity": 0.3,
              "width": 50,
            }
          }
        />
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "black",
              "borderRadius": 8,
              "height": 5,
              "marginHorizontal": 8,
              "opacity": 0.3,
              "width": 50,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "marginTop": 40,
        "paddingHorizontal": 20,
      }
    }
  >
    <View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={false}
      >
        <View
          style={
            Object {
              "backgroundColor": "#000000",
              "borderRadius": 8,
              "elevation": 5,
              "padding": 15,
              "shadowColor": "#000000",
              "shadowOffset": Object {
                "height": 3,
                "width": 0,
              },
              "shadowOpacity": 0.3,
              "shadowRadius": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Roboto-Regular",
                "fontSize": 18,
                "textAlign": "center",
              }
            }
          >
            Save
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
